Multi-EKF support

dagar opened the pull request:
**WORK IN PROGRESS**
Multi-EKF will be a focus of the v1.12 release cycle. I’m opening this early so we can begin to identify and address the architectural changes that will need to be in place to make this viable.


### Multi-EKF Background/Motivation
 WIP 

Currently in PX4 most commonly supported boards have 2 or 3 IMUs that are typically different models, sometimes with wildly different output data rates, and often on the same bus, or with no synchronization mechanism. As a result our sensor hub module isn't able to effectively compare raw sensor data and "vote", so it simply selects a highest priority sensor based on a simple set of fault metrics (timeouts, stale data, etc) and passes that on to the estimator. This allows the system to continue operating if a sensor is completely lost (a hard fault), but this is actually quite rare in operation. In reality most problems encountered in typical usage are soft faults that impact navigation like high vibration (aliasing, clipping) or erratic sensor problems that still produce output. 

To effectively utilize the current suite of heterogeneous IMUs to mitigate the most common real world pain points I'm proposing moving to an architecture of running (at least) one estimator per IMU. 

![Blank Diagram](https://user-images.githubusercontent.com/84712/79085293-024b8180-7d06-11ea-9db0-6c68cf0e59d6.png)

![Screenshot from 2020-04-12 13-38-07](https://user-images.githubusercontent.com/84712/79085940-62dbbe00-7d08-11ea-957d-db88ad433dc3.png)


 - in multi-mode each ekf2 instance publishes
        - `estimator_attitude` (vehicle_attitude)
        - `estimator_local_position` (vehicle_local_position)
        - `estimator_global_position` (vehicle_global_position)
 - a new `ekf2_selector` module consumes all estimator_* messages, selects a primary (latching), and republishes vehicle_attitude, vehicle_local_position, vehicle_global_position
 - `ekf2_selector` handles resets and deltas when switching between estimator instances
 - At the moment a switch will only occur if there's a fault while another estimator instance with a better combined test ratio (vel + pos + hgt) is fault free. Opportunity to identify (and store) the best sensor combination 
- sensors hub is now configurable to either provide the single highest priority IMU or publish all
   - new parameter `SENS_IMU_MODE` to configure IMU output
     - 0: publish all vehicle_imu instances
     - 1: publish only highest priority as primary
- On F7 we can start with one estimator instance per IMU. On F4 continue as is, user configurable if desired (lots of room for optimization on F4).



### TODO:
 - proper multi-module support (https://github.com/PX4/Firmware/pull/12191)
 - sensor_combined kill? (migrate to vehicle_imu)
 - ekf2 replay review (logging all raw data for all instances probably not a good default)
 - ecl analysis script updated to be aware of multiple instances
 - factor in initial sensor priorities (or kill this entire concept?)
     - the initial primary/preferred IMU should probably be configurable
 - preflight estimator_status (commander logic updated to only care about primary or should it check all?)
 - different origins?
     - each estimator instance will have a slightly different origin (or worse), does it matter or can we get away with it if resets between instances are carefully handled everywhere?
 - SITL add multiple sensor instances
 - sensor metadata should belong to each sensor instance, not `EKF2_*` parameters
      - offsets, rotations, possibly even noise
 - uORB multi-instance assumptions, we can't be sure all ekf2 publications are the same instance number. Maybe time to rethink uORB with namespaces? `estimator_status0 -> estimator/0/status`
 - lockstep implications (ekf2_timestamps)
 - Do we potentially need/want a separate level calibration per IMU?
 - per instance parameters (EKF2_MAGBIAS_{X, Y, Z}, accel and gyro noise?)
### Future
 - entire matrix of sensor combinations (more sensors module configuration)
 - combine onboard AND offboard estimators (UAVCAN v1)
 - testing ideas, run new and old changes side by side safely

jlecoeur commented:
Awesome! I like the proposed architecture very much. 

Some thoughts on the architecture:
Although more complex, you may consider moving the selected instance to the att_pos_ctl work queue for reduced latency and stack.
Thinking out loud, we may benefit from the EKF architecture (1 time-delayed EKF + 1 output filter) to optimize further. For example run 3 time-delayed EKFs in their own work queue, select one, and run the output filter for the selected instance only, potentially in the att_pos_ctl work queue to allow higher output rate and lower latency.

Regarding ORB instances, would not it be sufficient to add the ability to "--force" an instance number at the publish site?

dagar commented:
> Although more complex, you may consider moving the selected instance to the att_pos_ctl work queue for reduced latency and stack.

That's worth thinking about. I didn't get into it above, but `wq:att_pos_ctrl` is higher priority than `wq:INSx` and it's scheduled with publications of the current primary EKF (with backup timeout schedule). I was considering mechanisms for adjusting the priority of those threads dynamically based on EKF instance priority. With only 3 instances (1 per IMU) it's likely already fine as in many cases those sensors are on the same bus and already spaced out, however this will start to break down with a larger matrix instances. Even on something like an H7 where we could afford to run a full matrix of sensor combinations we'll still want to be careful to keep the latency to a minimum on the selected path.

> run the output filter for the selected instance only, potentially in the att_pos_ctl work queue to allow higher output rate and lower latency.

I'd like to explore that, and it's actually what I was trying to get at here https://github.com/PX4/Firmware/pull/14379#issuecomment-598385558. Instead of independent downsampling (hard coded filter update period) within each ecl/EKF backend we could simply update it with every new IMU. Configuring the IMU integration would be configuring the corresponding filter update period. We could update the attitude separately from buffered IMU (pre-integration).

> Regarding ORB instances, would not it be sufficient to add the ability to "--force" an instance number at the publish site?

The (potential) issue is something else coming in earlier and claiming that instance first. We can be careful to avoid that in practice, but I'd rather the potential hole didn't exist. Each ekf2 instance has up to 15 different publications. It would be a nice simplification if we can trivially map a particular message instance back to each estimator rather than every single message carrying metadata. I think this could be a good excuse to add a bit of hierarchy to uORB, but that alone might be a rabbit hole.

https://github.com/PX4/Firmware/blob/66eacd24bc8ba26f6d22a4596c16040e841d5dc3/src/modules/ekf2/ekf2_main.cpp#L269-L283

LorenzMeier commented:
Super nice to see this!

dagar commented:
The (potential) orb instance ordering issue has been resolved. With some uORB::Publication changes now in master at Ekf2 construction we advertise all topics at once.

xdwgood commented:
When performing EKF selection, how do we plan to solve the state jump?

dagar commented:
> When performing EKF selection, how do we plan to solve the state jump?

The selector handles the jump and we already have the reset mechanisms in place (messages and controllers). This is also used to pass through any actual resets from the active instance.

https://github.com/PX4/Firmware/blob/a4927606ed3799361819ec7d1caaf207fe6a269f/msg/vehicle_attitude.msg#L6

https://github.com/PX4/Firmware/blob/a4927606ed3799361819ec7d1caaf207fe6a269f/msg/vehicle_local_position.msg#L29

dagar commented:
![Screenshot from 2020-07-15 10-13-39](https://user-images.githubusercontent.com/84712/87555778-e5677700-c683-11ea-9286-b1bd615f110c.png)
![Screenshot from 2020-07-15 10-14-05](https://user-images.githubusercontent.com/84712/87555823-f0baa280-c683-11ea-84c7-e9e4ea120e90.png)
![Screenshot from 2020-07-15 10-14-31](https://user-images.githubusercontent.com/84712/87555887-016b1880-c684-11ea-9cf3-1ac38f1e3ab4.png)

dagar commented:
With the PX4/ecl move to Sympy we now have more than enough flash (fmu-v2, etc) to proceed with this.

The next blocker is https://github.com/PX4/Firmware/pull/15235 that splits mag out of `sensor_selection`, which will then allow the ekf2_selector to takeover IMU selection, while sensors continues selecting the primary mag.

priseborough commented:
@dagar I've started on some modifications to the selection logic here: https://github.com/priseborough/Firmware/commits/pr-ekf2_selector.

dagar commented:
Thanks @priseborough, I'm currently getting this into a form that's safe to merge for further incremental testing and development.

Everything is in place, but now disabled by default. To enable multi-EKF you have to configure the number of ekf2 instances (`param set EKF2_MULTI_INST 3`) and tell sensors to skip selecting the primary IMU (`param set SENS_IMU_MODE 0`). There's also initial support for specifying a particular mag to EKF2 instance (SENS_MAG_MODE + EKF2_x_MAG_ID), but that can be ignored for now.

dagar commented:
Currently my thinking is to get this safely into master (off by default), but once we have confidence in the selection logic to start enabling it by default roughly proportional to system resources.

 - STM32F4/K66: off by default
      - some specific setups can afford it, but memory is too tight on these older boards to enable by default everywhere
      - this could change if we use the same INS WQ thread and add priority mechanism to the WQ scheduling to make sure the primary always runs first immediately when IMU is available (to minimize latency) 
 - STM32F7: 1 instance per IMU, sensors hub still selects the primary mag/baro/gps
 - STM32H7/iMXRT/Linux/SITL: full matrix of sensor combinations (IMU x Mag)
      - gradually expanding to include all baros, GPS, and possibly airspeed
      - requires increasing the maximum number of ORB instances https://github.com/PX4/Firmware/pull/15647, https://github.com/PX4/Firmware/pull/15632, etc

dagar commented:
In an effort to move this forward I'm going to try and bring in everything separately up to the actual ekf2 selector + ekf2 multi instance change.

 - https://github.com/PX4/Firmware/pull/15656
 - https://github.com/PX4/Firmware/pull/15657
 - https://github.com/PX4/Firmware/pull/15658

priseborough commented:
Based on my testing of gyro step offset errors, we will probably need to add a gyro consistency check in the sensors module that flags when gyros disagree. That flag would then be used by the selector to switch more rapidly when this occurs. Otherwise achieving a timely switchover before the offset rolls the vehicle inverted is difficult to achieve if innovation test ratios are the only source of information.

priseborough commented:
@dagar are you able to run the latest version of this on SITL gazebo, ie 'make px4_sitl gazebo' ?

I am getting for following messages at the comnsole:

<img width="723" alt="Screen Shot 2020-09-04 at 10 32 05 am" src="https://user-images.githubusercontent.com/3596952/92186420-0be2ab00-ee9a-11ea-8e71-603e3b3da5e6.png">

dagar commented:
> Based on my testing of gyro step offset errors, we will probably need to add a gyro consistency check in the sensors module that flags when gyros disagree. That flag would then be used by the selector to switch more rapidly when this occurs. Otherwise achieving a timely switchover before the offset rolls the vehicle inverted is difficult to achieve if innovation test ratios are the only source of information.
> 
> Edit: After discussing this with @dagar, his suggestion was that could repurpose and rename the exisiting sensor pre-flight messages so that they publish the sensor inconsistency for each sensor ID wrt the primary for all operation. The estimator selection would then use that information to assist with fault detection and isolation.

See https://github.com/PX4/Firmware/pull/15681 for the accel & gyro inconsistency message repurposed to run beyond preflight and keep the data per sensor instance rather than only the maximum. One question is if the filtering is appropriate for both use cases.

``` C++
_gyro_diff[sensor_index] = 0.95f * _gyro_diff[sensor_index] + 0.05f * (primary_gyro - current_gyro);
```

priseborough commented:
@dagar Sorry - missed your updates https://github.com/PX4/Firmware/pull/15681 addressing the same issue.

priseborough commented:
Filtering ratio should be OK, but will be adjusted following testing.

dagar commented:
If anyone would like to give this a try.

``` Console
param set EKF2_MULTI_INST 3
param set SENS_IMU_MODE 0
reboot
```

dagar commented:
Here's a handheld test with a CUAV v5+ (3 IMUs). https://logs.px4.io/plot_app?log=40e25e03-2113-40bb-8600-15c5e9306de6
After an initial walk I manually stopped the icm20689, then the icm20602.

priseborough commented:
> * Would it make sense to more directly handle the offset between local position estimates (even though the deltas are already handled)?

That would introduce another layer of complexity and another mechanism whereby the fault instance being switched from would influence the offset applied to the healthy instance being switched to.

If there are three or more instances, then using median select algorithms for the local position and other state estimates starts to make sense.

priseborough commented:
> Here's a handheld test with a CUAV v5+ (3 IMUs). https://logs.px4.io/plot_app?log=40e25e03-2113-40bb-8600-15c5e9306de6
> 
> There are instances of rapid switching before actual sensor failure that need to be investigate.
> 

I ran a ground test using a CUAV V5 with multi ekf enabled, but without inducing IMU failures. There were no switch events, but on inspection of the log, only the first  instance logged non zero  errors scores and a healthy status, eg:

<img width="1279" alt="Screen Shot 2020-09-05 at 10 59 37 am" src="https://user-images.githubusercontent.com/3596952/92293446-eec7de00-ef66-11ea-958e-a5acda839b7c.png">

<img width="1284" alt="Screen Shot 2020-09-05 at 10 59 45 am" src="https://user-images.githubusercontent.com/3596952/92293448-f12a3800-ef66-11ea-9ff0-297f169e6807.png">

EKF2_MULTI_INST and SENS_IMU_MODE were set as per you testing

> param set EKF2_MULTI_INST 3  # enable up to 3 instances of EKF2
> param set SENS_IMU_MODE 0    # disable sensors module primary IMU selection (sensor_selection publication)

priseborough commented:
jmavsim SITL testing using the same version and param settings for EKF2_MULTI_INST and SENS_IMU_MODE did work as expected, eg:

<img width="1278" alt="Screen Shot 2020-09-05 at 11 19 02 am" src="https://user-images.githubusercontent.com/3596952/92293792-a4942c00-ef69-11ea-9c6a-31a0ddac315e.png">

<img width="1279" alt="Screen Shot 2020-09-05 at 11 19 11 am" src="https://user-images.githubusercontent.com/3596952/92293796-a65def80-ef69-11ea-854c-dccdc682dde0.png">

dagar commented:
> I ran a ground test using a CUAV V5 with multi ekf enabled, but without inducing IMU failures. There were no switch events, but on inspection of the log, only the first instance logged non zero errors scores and a healthy status, eg:
> 

There was a race condition the first time you start with `EKF2_MULTI_INST` set because it goes through the available IMUs (published vehicle_imu instances) and distributes them per EKF2 instance (setting EKF2_x_IMU_ID). The problem was combining that with the newer IMU drivers that have relatively long initial resets (30-100 ms) following the datasheet to the letter. After the initial assignment the next boot skips the scan and starts all instances immediately. We'll have to keep an eye on it to see if it's resolved.

To inspect the system to see if it's working before waiting for a log you can check `ekf2 status`, `work_queue status`, or listener `estimator_selector_status`.


``` Console
nsh> ekf2 status
INFO  [ekf2] available instances: 3
INFO  [ekf2] 0: ACC: 3670026, GYRO: 3670026, MAG: 396825, healthy, test ratio: 0.00004 (0.00000) *
INFO  [ekf2] 1: ACC: 3932170, GYRO: 3932170, MAG: 396825, healthy, test ratio: 0.00015 (1.00000)
INFO  [ekf2] 2: ACC: 4259850, GYRO: 4325386, MAG: 396825, healthy, test ratio: 0.00018 (0.97297)

EKF2 Instance: 0
INFO  [ekf2] attitude:        valid
INFO  [ekf2] local position:  invalid
INFO  [ekf2] global position: invalid
ekf2: update: 85544 events, 18317625us elapsed, 214.13us avg, min 4us max 1180us 220.590us rms

EKF2 Instance: 1
INFO  [ekf2] attitude:        valid
INFO  [ekf2] local position:  invalid
INFO  [ekf2] global position: invalid
ekf2: update: 85446 events, 23057793us elapsed, 269.85us avg, min 4us max 2301us 330.451us rms

EKF2 Instance: 2
INFO  [ekf2] attitude:        valid
INFO  [ekf2] local position:  invalid
INFO  [ekf2] global position: invalid
ekf2: update: 85381 events, 24824014us elapsed, 290.74us avg, min 4us max 3338us 383.956us rms
```

priseborough commented:
After repeating the calibration of the IMU to set the sensor ID's correctly, I was able to replicate the switching behaviour @dagar encountered - here is the log https://logs.px4.io/plot_app?log=7be12c1b-d30b-4ae1-bf86-2c444ab8291c

The unwanted switching events appear to be due to incorrect calculation of the relative error scores, eg:

Absolute errors:

<img width="1286" alt="Screen Shot 2020-09-05 at 1 27 55 pm" src="https://user-images.githubusercontent.com/3596952/92296807-ceeee500-ef7b-11ea-8d2f-c1d832b65c4a.png">

Relative errors:

<img width="1286" alt="Screen Shot 2020-09-05 at 1 28 14 pm" src="https://user-images.githubusercontent.com/3596952/92296809-d615f300-ef7b-11ea-8dfd-f60dce5c96c4.png">

priseborough commented:
I've pushed a commit that fixes the unwanted switching behaviour. We may need to make _err_reduce_thresh adjustable via the para system. See following relative error scores and inatance selected index.

<img width="1279" alt="Screen Shot 2020-09-05 at 1 51 31 pm" src="https://user-images.githubusercontent.com/3596952/92297091-ebd8e780-ef7e-11ea-96a5-6522b0da1673.png">

The next TODO items for me are:

1) Rebase when https://github.com/PX4/Firmware/pull/15681 is merged and add the logic to utilise the IMU difference data
2) Add a convenient way to induce IMU errors for SITL testing

dagar commented:
> The next TODO items for me are:
> 
> 2. Add a convenient way to induce IMU errors for SITL testing

@priseborough do you have a pretty good idea of everything you'd like to see here? 
  - add bias, additional noise, etc
  - static from parameter, intermittent, or injected randomly in flight?

@julianoes has an ongoing effort for failure injection. https://github.com/PX4/Firmware/pull/15063

dagar commented:
> After repeating the calibration of the IMU to set the sensor ID's correctly, 

That's weird, the ID is from the sensor itself so all of this setup should function the same with or without calibrations. If it happens again please keep the full log and we'll debug. You can also reset the per EKF2 instance IMU IDs to force it to setup again (`EKF2_x_IMU_ID`).

The entire automatic setup and assignment mechanism will need to be reworked when this expands to include mags, baros, etc.

dagar commented:
@priseborough https://github.com/PX4/Firmware/pull/15681 has merged and I've rebased the `pr-ekf2_selector` branch on current master.

The gyro inconsistencies are at your disposal in the selector. You can find the device ids of sensors in use by each instance in estimator status (accel_device_id/gyro_device_id/baro_device_id/mag_device_id).

https://github.com/PX4/Firmware/blob/a257e0ead9d2f5a4b7ff9fe140fe278c8217e646/src/modules/ekf2/EKF2Selector.cpp#L246

dagar commented:
Rebased again to resolve a minor conflict with master and in the process I've squashed the EKF2 selector and fixes into a single commit + sensors module changes for support (https://github.com/PX4/Firmware/pull/15658).

dagar commented:
> We may need to make _err_reduce_thresh adjustable via the para system. 

I've added `EKF2_SEL_ERR_RED` to replace the `_err_reduce_thresh` constant. It will only be active in multi-ekf2 mode when the selector is running. https://github.com/PX4/Firmware/pull/14650/commits/bfb3dc833ad57fea586b23f410e063e25be726e5

priseborough commented:
I've prototyped checking of IMU gyro differences here: https://github.com/priseborough/Firmware/commit/5be878502258799567f0fd9f91a1f1a10c32233e

The next step is to try and use https://github.com/PX4/Firmware/pull/15063 for testing and tuning it.

dagar commented:
The last supporting PR (https://github.com/PX4/Firmware/pull/15655) has merged. I'll give this branch a final rebase and from this point forward we can add new commits, merge master, and then finally squash + merge once ready.

dagar commented on src/modules/ekf2/EKF2Selector.cpp:
```diff
@@ -241,43 +242,77 @@ void EKF2Selector::Run()
+					if (sensors_status_imu.gyro_device_ids[i] == sensors_status_imu.gyro_device_id_primary) {
+						_accumulated_gyro_error[i] = 0.0f;
+					} else {
+						_accumulated_gyro_error[i] += (sensors_status_imu.gyro_inconsistency_rad_s[i] - angle_rate_threshold) * time_step_s;
```
Should this be limited to gyros the sensors module has deemed healthy (no timeouts, etc)?

priseborough replied:
Are you talking about an error that would change primary instance? In such an event, then the accumulated error scores may need to be reset, depending on how the ekf selector handles the change in primary instance. I need to discuss this with you on a call.

dagar commented:
The failure injection PR (https://github.com/PX4/Firmware/pull/15063) has been merged.

dagar commented on src/modules/ekf2/EKF2Selector.cpp:
```diff
@@ -241,43 +242,77 @@ void EKF2Selector::Run()
 	for (int i = 0; i < _available_instances; i++) {
+		// if the gyro used by the EKF is faulty, declare the EKF unhealthy without delay
+		if (_gyro_fault_detected && _instance[i].estimator_status.gyro_device_id == faulty_gyro_id) {
+			_instance[i].healthy = false;
```
This might be better consolidated within `updateErrorScores()` so that there's only a single place making the healthy determination.

priseborough commented:
I've reworked the sensor differencing reporting, ekf2 selector checks that consume the data and added checking of accelerometers. Testing has been limited to setting calibration offsets on a V5 board to simulate sensor failures.

priseborough commented:
Test output generated using a 3 IMU V5 board with the first IMU X gyro bias set to 0.3 r/s to cause a switch:

Sensor status message changes as expected:

![Screen Shot 2020-09-18 at 1 39 30 pm](https://user-images.githubusercontent.com/3596952/93552764-bbe50780-f9b4-11ea-9462-893356d6902d.png)

Estimator selector accumulated gyro error increments as expected with selection change induced by innovation failure before the accumulated angular difference gyro fault threshold was reached:
![Screen Shot 2020-09-18 at 1 50 28 pm](https://user-images.githubusercontent.com/3596952/93553269-f0a58e80-f9b5-11ea-81fd-10d01637aee9.png)

TODO: Tuning of the parameters and further testing

dagar commented on src/modules/sensors/voted_sensors_update.cpp:
```diff
@@ -405,32 +405,48 @@ void VotedSensorsUpdate::setRelativeTimestamps(sensor_combined_s &raw)
-
-			const Vector3f current_accel{_last_sensor_data[sensor_index].accelerometer_m_s2};
-			_accel_diff[sensor_index] = 0.95f * _accel_diff[sensor_index] + 0.05f * (primary_accel - current_accel);
+		if (_accel_device_id[sensor_index] != 0) {
```
```suggestion
		if ((_accel_device_id[sensor_index] != 0) && (_accel.priority[sensor_index] > 0)) {
```

Priority 0 means sensor is disabled.

dagar replied:
Done.

priseborough commented:
I forked and added support for a third IMU to assist with testing here : https://github.com/priseborough/Firmware/tree/pr-ekf2_selector

dagar commented:
> I forked and added support for a third IMU to assist with testing here : https://github.com/priseborough/Firmware/tree/pr-ekf2_selector

I could add that properly for SITL with a parameter to select the number of IMUs/mags/etc.

dagar commented:
I've reworked and simplified the ekf2 multi startup and configuration. Instead of each ekf2 instance having a corresponding parameter for IMU and mag device id you now configure the maximum number of IMUs (EKF2_MULTI_IMU) and mags (EKF2_MULTI_MAG) to use. The startup code (ekf2_main) continues running in the background for 60 seconds until all desired instances are allocated or the vehicle arms. I've also enabled it by default in SITL for testing.

```
SENS_IMU_MODE 0 # required for multi-ekf
EKF2_MULTI_IMU 3 # ideally set to the number of IMUs actually in the system

SENS_MAG_MODE 0 # if you want ekf instances per mag
EKF2_MULTI_MAG # set to the max number of mags you want EKF2 instances for
```

dagar commented:
For testing at this stage I've enabled multi-EKF by default in SITL, px4_fmu-v5, and holybro_durandal.

dagar commented:
Next steps
 - fix SITL lockstep with multi-ekf
 - finish ecl analysis multi-ekf support
 - flight testing

dagar commented:
Multi-ekf is now working with lockstep and the ecl analysis script will now produce reports for all estimator instances in a ulg.

dagar commented:
``` Console

nsh> ekf2 status
INFO  [ekf2] available instances: 6
INFO  [ekf2] 0: ACC: 3670026, GYRO: 3670026, MAG: 396825, healthy, test ratio: 0.0001771 (0.00000) *
INFO  [ekf2] 1: ACC: 3670026, GYRO: 3670026, MAG: 396809, healthy, test ratio: 0.0010318 (1.00000) 
INFO  [ekf2] 2: ACC: 4259850, GYRO: 4325386, MAG: 396809, healthy, test ratio: 0.0004816 (1.00000) 
INFO  [ekf2] 3: ACC: 3932170, GYRO: 3932170, MAG: 396825, healthy, test ratio: 0.0006842 (1.00000) 
INFO  [ekf2] 4: ACC: 4259850, GYRO: 4325386, MAG: 396825, healthy, test ratio: 0.0008552 (1.00000) 
INFO  [ekf2] 5: ACC: 3932170, GYRO: 3932170, MAG: 396809, healthy, test ratio: 0.0008783 (1.00000) 

nsh> listener estimator_attitude

TOPIC: estimator_attitude 6 instances

Instance 0:
 vehicle_attitude_s
        timestamp: 166258406  (0.002771 seconds ago)
        timestamp_sample: 166257915  (491 us before timestamp)
        q: [-0.3600, -0.0237, -0.0173, 0.9324]  (Roll: -0.8 deg, Pitch: 3.3 deg, Yaw: -137.8 deg)
        delta_q_reset: [1.0000, 0.0000, -0.0000, 0.0015]  (Roll: 0.0 deg, Pitch: -0.0 deg, Yaw: 0.1 deg)
        quat_reset_counter: 2

Instance 1:
 vehicle_attitude_s
        timestamp: 166263859  (0.002006 seconds ago)
        timestamp_sample: 166262914  (945 us before timestamp)
        q: [0.3477, 0.0220, 0.0103, -0.9372]  (Roll: -0.2 deg, Pitch: 2.8 deg, Yaw: -139.3 deg)
        delta_q_reset: [1.0000, 0.0000, 0.0000, -0.0005]  (Roll: 0.0 deg, Pitch: 0.0 deg, Yaw: -0.1 deg)
        quat_reset_counter: 2

Instance 2:
 vehicle_attitude_s
        timestamp: 166266762  (0.000865 seconds ago)
        timestamp_sample: 166266605  (157 us before timestamp)
        q: [0.3498, 0.0309, 0.0131, -0.9362]  (Roll: -0.1 deg, Pitch: 3.9 deg, Yaw: -139.0 deg)
        delta_q_reset: [1.0000, -0.0000, 0.0000, -0.0002]  (Roll: -0.0 deg, Pitch: 0.0 deg, Yaw: -0.0 deg)
        quat_reset_counter: 2

Instance 3:
 vehicle_attitude_s
        timestamp: 166267979  (0.002977 seconds ago)
        timestamp_sample: 166267665  (314 us before timestamp)
        q: [-0.6703, 0.0192, -0.0212, -0.7414]  (Roll: 0.3 deg, Pitch: 3.3 deg, Yaw: 95.8 deg)
        delta_q_reset: [1.0000, -0.0000, -0.0000, 0.0015]  (Roll: -0.0 deg, Pitch: -0.0 deg, Yaw: 0.1 deg)
        quat_reset_counter: 2

Instance 4:
 vehicle_attitude_s
        timestamp: 166272378  (0.003100 seconds ago)
        timestamp_sample: 166271621  (757 us before timestamp)
        q: [-0.6769, 0.0213, -0.0231, -0.7353]  (Roll: 0.2 deg, Pitch: 3.6 deg, Yaw: 94.7 deg)
        delta_q_reset: [1.0000, -0.0000, 0.0000, 0.0012]  (Roll: -0.0 deg, Pitch: 0.0 deg, Yaw: 0.1 deg)
        quat_reset_counter: 2

Instance 5:
 vehicle_attitude_s
        timestamp: 166323060  (0.004335 seconds ago)
        timestamp_sample: 166322697  (363 us before timestamp)
        q: [0.3499, 0.0289, 0.0121, -0.9362]  (Roll: -0.1 deg, Pitch: 3.6 deg, Yaw: -139.0 deg)
        delta_q_reset: [1.0000, 0.0000, -0.0000, -0.0002]  (Roll: 0.0 deg, Pitch: -0.0 deg, Yaw: -0.0 deg)
        quat_reset_counter: 2

```

priseborough commented:
I've encountered difficulty with using the MAVSDK test environment to test IMU faults (for example it seems to be failing all IMU's) and have stopped work on it until I can set up a call with someone that knows how it works.

dagar commented:
> I've encountered difficulty with using the MAVSDK test environment to test IMU faults (for example it seems to be failing all IMU's) and have stopped work on it until I can set up a call with someone that knows how it works.

@priseborough after merging https://github.com/PX4/Firmware/pull/15900.

Set `SYS_FAILURE_EN` 1,

Example: fail the first accel

![Screenshot from 2020-10-07 10-57-48](https://user-images.githubusercontent.com/84712/95348378-f9d69c80-088b-11eb-9dd2-d36eb54f3dd4.png)

dagar commented:
From my perspective the short term goal from this point is getting this to a safe, mergeable state with the selector off by default. In parallel we can proceed with testing and iteration of the ekf2 selection logic with testers opting in. With that in mind, @bresch could you review the core EKF2 changes? You're also more than welcome to review beyond that as well.

dagar commented:
We've hit the fmu-v2 flash limit again.

![Screenshot from 2020-10-19 11-48-50](https://user-images.githubusercontent.com/84712/96474729-1b774280-1201-11eb-8d93-d283e1d2cbb2.png)

dagar commented:
Real test logs thanks to @mcsauder.

https://review.px4.io/plot_app?log=f21e4e44-8d77-4758-b2db-091669e7bf71
https://review.px4.io/plot_app?log=1898560b-867f-4c2a-bd4d-5f5ccc9013a0

In the second log you can see an estimator switch midflight with a large uncommanded change in yaw setpoint.

<img width="1287" alt="Screen Shot 2020-10-20 at 11 21 34 PM" src="https://user-images.githubusercontent.com/84712/96672855-321db680-1333-11eb-95fe-84b90a40bf9a.png">

I don't see a problem with the logic in the estimator selector (the delta isn't wrong), I suspect this is a problem with the existing reset logic in the muticopter attitude controller (mc_att_control) and stabilized mode.

https://github.com/PX4/Firmware/blob/0b74076265edda2ac34cb9c4addcfeea7277d3a2/src/modules/mc_att_control/mc_att_control_main.cpp#L253-L261

dagar commented:
I've fixed the delta_q_reset in the EKF2 selector.  https://github.com/PX4/Firmware/pull/14650/commits/e336daefb2626f8729f4a6a1396a0738a791b56c

priseborough reviewed (approved):
This is off by default so safe to merge. Items to look at post merge include:

MAVSDK needs to be updated to support the failure of a single sensor instance and auto tests updated.
EKF replay with multiple instances if possible. Single instance appears to work.

dagar commented:
Thanks Paul.

Next steps from my perspective.
 - multi-EKF enabled in SITL (lockstep issues to resolve)
 - MAVSDK tests with specific sensors (failure, inject bias, noise, etc)
 - purge sensor_combined entirely
 - ekf2 replay
     - multi-ekf support
     - working with vehicle_imu
     - possibly remove ekf2_timestamps?
 - multi module support generalize in ModuleBase if possible and remove custom EKF2 handling
 - extend architecture to support multi-ekf across any type of sensor (barometers, GPS, optical flow, range finder, etc)
 - flight testing on common boards (2-3 IMUs, 1 internal, 1 or more external mags)

mrpollo commented:
TODO:
* Make sure we highlight this feature on the next release notes
* Create documentation on how the feature works, and how to configure it
